omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -92,55 +100,158 @@ pub struct PersistentStorage<T>
 where
     T: Entry,
 {
-    /// Disk-based commit log for entries
-    commitlog: CommitLog,
-    /// The path to the directory containing a commitlog
-    log_path: String,
-    /// Local sled key-value store, enabled by default
-    sled: Db,
+    /// Local RocksDB key-value store
+    db: DB,
+    /// Buffered, atomic write batch
+    write_batch: WriteBatchWithTransaction<false>,
+    /// The index of the next log entry to be appended
+    next_log_key: usize,
     /// A placeholder for the T: Entry
     t: PhantomData<T>,
 }
 
-impl<T: Entry> PersistentStorage<T> {
+impl<T: Entry> PersistentStorage<T>
+where
+    T: Entry + Serialize + for<'a> Deserialize<'a>,
+    T::Snapshot: Serialize + for<'a> Deserialize<'a>,
+{
     /// Creates or opens an existing storage
     pub fn open(storage_config: PersistentStorageConfig) -> Self {
-        let path = storage_config.path.expect("No path found in config");
-
-        let commitlog =
-            CommitLog::new(storage_config.commitlog_options).expect("Failed to create Commitlog");
+        // Create database with log column
+        let path = storage_config.path;
+        let log_cf = ColumnFamilyDescriptor::new(LOG, storage_config.log_options);
+        let db_opts = storage_config.rocksdb_options;
+        let db = rocksdb::DB::open_cf_descriptors(&db_opts, path, vec![log_cf])
+            .expect("Failed to create RocksDB");
+        let log_handle = db
+            .cf_handle(LOG)
+            .expect("Failed to create RocksDB log column family");
+
+        // Create next log key from the current max log key
+        let mut log_iter = db.raw_iterator_cf(log_handle);
+        log_iter.seek_to_last();
+        let next_log_key = if log_iter.valid() {
+            let key = log_iter.key().unwrap();
+            assert_eq!(
+                key.len(),
+                8,
+                "Couldn't recover storage: Log key has unexpected format."
+            );
+            usize::from_be_bytes([
+                key[0], key[1], key[2], key[3], key[4], key[5], key[6], key[7],
+            ]) + 1
+        } else {
+            match db
+                .get(TRIM)
+                .expect("Couldn't recover storage: Reading compacted_idx failed.")
+            {
+                Some(bytes) => usize::read_from(bytes.as_bytes())
+                    .expect("Couldn't recover storage: Commpacted index has unexpected format."),
+                None => 0,
+            }
+        };
+        drop(log_iter);
```

Comment:
Add comments for what we're doing here. I guess we are reading and deserializing 8 bytes that corresponds to the key?